### PR TITLE
[RLlib] Move action distributions out of RLModule forwards

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

Since eager tracing and torch compile don't like convoluted python logic inside their traced functions,
having action distributions is there breaks graphs.

This PR moves action distributions out of forward methods and introduces a method "get_action_dist_cls" to RL Modules
that can be called from outside RLModules to construct the action distribution from the outputs of the forward methods.

### CODE REVIEW

**Review by kouroshHakha — COMMENTED** [review]
Overall looks good. I have one big input after reading the whole PR.

We need to still be able to independently indicate the distribution class type for each phase of RL training (e.g. inference vs. exploration vs. training). My recommendation is to first define the new API as part of the RLModule's public API. 
The requirements of this new API:
- Have it take a mode: RLMode and then allow users to have more flexible distribution type selection. 

```
if explore:
    dist = self.model.get_action_distribution_cls(RLMode.Exploration)
else:
    dist = self.model.get_action_distribution_cls(RLMode.Inference)
```

- It should check that the user always returns the expected type (TorchDistribution or TfDistribution)

In BCModule if I return torch.distributions.Categorical it should yell at me for not using TorchDistribution

File: rllib/core/testing/torch/bc_module.py [hunk 1]
```diff
@@ -84,6 +89,9 @@ def __init__(
             nn.Linear(hidden_dim, action_dim),
         )
 
+    def get_action_dist_cls(self):
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/core/testing/torch/bc_module.py:92]: Why? This is imported with`from ray.rllib.models.torch.torch_distributions import TorchCategorical`

File: rllib/core/rl_module/rl_module.py [hunk 2]
```diff
@@ -277,6 +277,13 @@ class RLModule(abc.ABC):
 
     framework: str = None
 
+    # Constants that signify forward pass modes.
```
Comment (kouroshHakha) [inline rllib/core/rl_module/rl_module.py:280]: enum these plz.

File: rllib/core/rl_module/rl_module.py [hunk 3]
```diff
@@ -329,6 +336,28 @@ def setup(self):
         abstraction can be used to create any component that your RLModule needs.
         """
 
+    def get_action_dist_cls(self, mode: str) -> Type[Distribution]:
```
Comment (kouroshHakha) [inline rllib/core/rl_module/rl_module.py:339]: type hint: mode: RLMode  (Enum)

File: rllib/core/rl_module/rl_module.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/core/rl_module/rl_module.py:280]: and don't forget to update the docstrings accordingly.

File: rllib/core/testing/torch/bc_learner.py [hunk 4]
```diff
@@ -12,6 +12,10 @@ def compute_loss_per_module(
         self, module_id: str, batch: SampleBatch, fwd_out: Mapping[str, TensorType]
     ) -> Mapping[str, Any]:
 
-        action_dist = fwd_out["action_dist"]
-        loss = -torch.mean(action_dist.log_prob(batch[SampleBatch.ACTIONS]))
+        action_dist_inputs = fwd_out[SampleBatch.ACTION_DIST_INPUTS]
+        action_dist_class = self._module[module_id].get_action_dist_cls(
+            self._module[module_id].TRAIN
```
Comment (kouroshHakha) [inline rllib/core/testing/torch/bc_learner.py:17]: use Enum

File: rllib/core/rl_module/torch/torch_rl_module.py [hunk 5]
```diff
@@ -54,6 +55,9 @@ def __init__(self, *args, **kwargs) -> None:
         # the interface of that base-class not the actual implementation.
         self.config = self.unwrapped().config
 
+    def get_action_dist_cls(self, *args, **kwargs) -> Type[TorchDistribution]:
```
Comment (kouroshHakha) [inline rllib/core/rl_module/torch/torch_rl_module.py:58]: does it have to be *args, **kwargs? Why not inherit the signature?

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py [hunk 6]
```diff
@@ -39,14 +39,23 @@ def compute_loss_per_module(
         # learning rate for that agent.
         # TODO (Kourosh): come back to RNNs later
 
-        curr_action_dist = fwd_out[SampleBatch.ACTION_DIST]
-        action_dist_class = type(fwd_out[SampleBatch.ACTION_DIST])
-        prev_action_dist = action_dist_class.from_logits(
+        action_dist_class_train = self._module[module_id].get_action_dist_cls(
+            self._module[module_id].TRAIN
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/torch/ppo_torch_learner.py:43]: use the Enum

File: rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py
Comment (kouroshHakha) [inline rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py:1]: We got rid of all policy_rlm files. Why are they showing up here?

File: rllib/algorithms/ppo/torch/ppo_torch_policy_rlm.py
Comment (kouroshHakha) [inline rllib/algorithms/ppo/torch/ppo_torch_policy_rlm.py:1]: We got rid of these files. They should not show up here.

File: rllib/algorithms/appo/tf/appo_tf_policy_rlm.py
Comment (kouroshHakha) [inline rllib/algorithms/appo/tf/appo_tf_policy_rlm.py:1]: We got rid of this file before.

File: rllib/algorithms/appo/torch/appo_torch_policy_rlm.py
Comment (kouroshHakha) [inline rllib/algorithms/appo/torch/appo_torch_policy_rlm.py:1]: We got rid of this file as well.

File: rllib/algorithms/impala/tf/impala_tf_policy_rlm.py
Comment (kouroshHakha) [inline rllib/algorithms/impala/tf/impala_tf_policy_rlm.py:1]: extra file

File: rllib/algorithms/impala/torch/impala_torch_policy_rlm.py
Comment (kouroshHakha) [inline rllib/algorithms/impala/torch/impala_torch_policy_rlm.py:1]: extra file

File: rllib/algorithms/ppo/tests/test_ppo_rl_module.py [hunk 7]
```diff
@@ -47,7 +47,7 @@ def get_expected_module_config(
     return config
 
 
-def dummy_torch_ppo_loss(batch, fwd_out):
+def dummy_torch_ppo_loss(module, batch, fwd_out):
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/tests/test_ppo_rl_module.py:50]: did we not get rid of this at some point?

**Review by kouroshHakha — COMMENTED** [review]
Two things:

- I think merge conflicts are not properly resolved leaving traces of things that have been deleted. Please double check with master. 
- Use Enum. It will make the whole PR nicer and less verbose when you want to pass the mode into the the get_action_dist API. instead of 

```
dist = self.modules[id].get_action_dist_class(self.modules[id].TRAIN)
```

it can be 

```
dist = self.modules[id].get_action_dist_class(RLMode.TRAIN)
```

File: rllib/core/rl_module/rl_module.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/rl_module/rl_module.py:280]: As per our offline discussion, the getter will be split up between three get methods and thus eliminate the need for the enumeration. The tradeoff being that if one wants to inherit from RLModule, they must implement three methods but there is no need for such constants to be used when getting action dist classes.

File: rllib/core/rl_module/torch/torch_rl_module.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/rl_module/torch/torch_rl_module.py:58]: Passing through arguments makes it easier to hack the underlying get_action_dist_cls method, that's it.
The tradeoff is the missing signature.

File: rllib/algorithms/ppo/tests/test_ppo_rl_module.py (hunk 7 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/ppo/tests/test_ppo_rl_module.py:50]: I can't recall us getting rid of it and it's still on master.

File: doc/source/rllib/rllib-rlmodule.rst [hunk 8]
```diff
@@ -426,7 +426,7 @@ What your customization could have looked like before:
...
+All of the ``Policy.compute_***`` functions expect that `~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` and `~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference` return a dictionary that contains the key "action_dist_inputs" mapping to inputs to a ``ray.rllib.models.distributions.Distribution`` class. Commonly used distribution implementations can be found under ``ray.rllib.models.tf.tf_distributions`` for tensorflow and ``ray.rllib.models.torch.torch_distributions`` for torch. You can choose to return determinstic actions, by creating a determinstic distribution instance. See `Writing Custom Single Agent RL Modules`_ for more details on how to implement your own custom RL Module.
```
Comment (sven1977) [inline doc/source/rllib/rllib-rlmodule.rst:429]:
```suggestion
All of the ``Policy.compute_***`` functions expect that `~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` and `~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference` return a dictionary that contains the key "action_dist_inputs", whose value are the parameters (inputs) of a ``ray.rllib.models.distributions.Distribution`` class. Commonly used distribution implementations can be found under ``ray.rllib.models.tf.tf_distributions`` for tensorflow and ``ray.rllib.models.torch.torch_distributions`` for torch. You can choose to return determinstic actions, by creating a determinstic distribution instance. See `Writing Custom Single Agent RL Modules`_ for more details on how to implement your own custom RL Module.
```

**Review by sven1977 — APPROVED** [review]
Thanks for this PR @ArturNiederfahrenhorst !
LGTM now, let's merge, it will be unblocking https://github.com/ray-project/ray/pull/34959, which also requires distributions objects no longer being returned by forward calls.

**Review by sven1977 — APPROVED** [review]
Thanks for this PR @ArturNiederfahrenhorst !
LGTM now, let's merge, it will be unblocking https://github.com/ray-project/ray/pull/34959, which also requires distributions objects no longer being returned by forward calls.

**Pull request merged**